streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordQueueTest.java

```diff
@@ -136,4 +138,24 @@ public void shouldThrowStreamsExceptionWhenValueDeserializationFails() throws Ex
 
         queue.addRawRecords(records, timestampExtractor);
     }
+
+    @Test(expected = StreamsException.class)
+    public void shouldThrowOnNegativeTimestamp() {
+        final byte[] value = Serdes.Long().serializer().serialize("foo", 1L);
+        final List<ConsumerRecord<byte[], byte[]>> records = Collections.singletonList(
+                new ConsumerRecord<>("topic", 1, 1, -1L, TimestampType.CREATE_TIME, 0L, 0, 0, recordKey, recordValue));
+
+        queue.addRawRecords(records, new FailOnInvalidTimestamp());
+    }
+
+    @Test
+    public void shouldDropOnNegativeTimestamp() {
+        final byte[] value = Serdes.Long().serializer().serialize("foo", 1L);
+        final List<ConsumerRecord<byte[], byte[]>> records = Collections.singletonList(
+                new ConsumerRecord<>("topic", 1, 1, -1L, TimestampType.CREATE_TIME, 0L, 0, 0, recordKey, recordValue));
+
+        queue.addRawRecords(records, new LogAndSkipOnInvalidTimestamp());
```

Comment:
We have thought about different extractors per topic, yes, but even with that I think we can keep a field per `RecordQueue`, so that is not a big deal really and we can consider changing it accordingly if necessary.